### PR TITLE
Add fips-3678 feature

### PR DESCRIPTION
**Pull request by behrat**
This PR contains multiple logical commits. The ones before "Add fips-3678" I think can actually stand on their own, and are useful even without FIPS support.

Since some of the APIs are not available in the FIPS-validated version, we must add some conditional-compilation in order to successfully link against the FIPS-validated version of BoringSSL (see the addition in `README.md` for details). This can be controlled with the new `fips-3678` feature. The intention of naming the feature this way is that future FIPS-validated versions of BoringSSL can by supported by adding a `fips-xxxx` feature, while still supporting the `fips-3678` for backwards-compatability.

### CODE REVIEW

File: boring-sys/build.rs [hunk 1]
```diff
@@ -176,24 +176,25 @@ fn main() {
...
-                "submodule",
-                "update",
-                "--init",
-                "--recursive",
-                "deps/boringssl",
-            ])
-            .status();
-        if !status.map_or(false, |status| status.success()) {
-            panic!("failed to fetch submodule - consider running `git submodule update --init --recursive deps/boringssl` yourself");
+    println!("cargo:rerun-if-env-changed=BORING_BSSL_PATH");
```
Comment (jyn514) [inline boring-sys/build.rs:179]: Hmm, shouldn't we have this for all the other env variables too? Off the top of my head I see `OPT_LEVEL`, `DEBUG`, `CARGO_CFG_TARGET_ARCH`, `CARGO_CFG_TARGET_OS`, ... https://doc.rust-lang.org/cargo/reference/build-scripts.html#rerun-if-env-changed says that env variables set by cargo are automatically rerun, but I suspect there's still a few that aren't being caught.

Happy to do that in a separate PR to avoid making this one too big.

File: boring/src/x509/mod.rs [hunk 2]
```diff
@@ -493,18 +493,18 @@ impl X509Ref {
     /// Returns the certificate's Not After validity period.
     pub fn not_after(&self) -> &Asn1TimeRef {
         unsafe {
-            let date = X509_getm_notAfter(self.as_ptr());
+            let date = X509_get0_notAfter(self.as_ptr());
```
Comment (jyn514) [inline boring/src/x509/mod.rs:511]: > since we're returning an unmutable reference, it's not possible to call any mutable functions of Asn1TimeRef (without using unsafe again): https://docs.rs/boring/1.1.6/boring/asn1/struct.Asn1TimeRef.html Another way to say it: returning an unmutable reference is effectively propagating the same guarantee as *const. (right?)

Doesn't an immutable reference guarantee that it won't be changed by this code *or any other code*? The docs say

> X509_getm_notBefore() and X509_getm_notAfter() are similar to X509_get0_notBefore() and X509_get0_notAfter() except they return non-constant mutable references

which seems to imply that these could change because the OpenSSL library itself decides to change them. I don't know if that's valid while another piece of code has an immutable reference to them.

File: boring-sys/build.rs [hunk 3]
```diff
@@ -205,9 +205,12 @@ fn main() {
         cfg.build_target("bssl").build().display().to_string()
     });
 
-    let build_path = get_boringssl_platform_output_path();
-    let build_dir = format!("{}/build/{}", bssl_dir, build_path);
-    println!("cargo:rustc-link-search=native={}", build_dir);
+    println!("cargo:rerun-if-env-changed=BORING_BSSL_LIB_PATH");
+    let lib_path = std::env::var("BORING_BSSL_LIB_PATH").unwrap_or_else(|_| {
```
Comment (jyn514) [inline boring-sys/build.rs:209]: Why this change? What are you using it for?

File: README.md [hunk 4]
```diff
@@ -22,6 +22,27 @@ _Notes_: The crate will look for headers in the `$BORING_BSSL_INCLUDE_PATH/opens
...
+3678](https://csrc.nist.gov/projects/cryptographic-module-validation-program/certificate/3678)
+is supported by this crate. Support is enabled by this crate's `fips-3678` feature.
+
+You must provide and build your own FIPS-validated module in order for this
+crate to link against it (see the instalation instructions in the
+[BoringCrypto FIPS 140-2 Non-Proprietary Security Policy] (https://csrc.nist.gov/CSRC/media/projects/cryptographic-module-validation-program/documents/security-policies/140sp3678.pdf)
+).
+
+After building the FIPS-validated module, you must build this crate with the
+`BORING_BSSL_PATH`, `BORING_BSSL_INCLUDE_PATH`, and maybe the
```
Comment (jyn514) [inline README.md:38]: Why is this true? What happens if you don't set these variables?

File: boring/src/x509/mod.rs [hunk 5]
```diff
@@ -229,16 +228,28 @@ impl X509Builder {
...
     pub fn set_not_after(&mut self, not_after: &Asn1TimeRef) -> Result<(), ErrorStack> {
-        unsafe { cvt(X509_set1_notAfter(self.0.as_ptr(), not_after.as_ptr())).map(|_| ()) }
+        #[cfg(feature = "fips")]
+        unsafe {
+            cvt(X509_set_notAfter(self.0.as_ptr(), not_after.as_ptr())).map(|_| ())
+        }
+        #[cfg(not(feature = "fips"))]
+        unsafe {
+            cvt(X509_set1_notAfter(self.0.as_ptr(), not_after.as_ptr())).map(|_| ())
+        }
```
Comment (jyn514) [inline boring/src/x509/mod.rs:240]: Is `set_notAfter` still present in the latest version? Can we use that unconditionally instead of adding two different paths?

**Review by jyn514 — CHANGES_REQUESTED** [review]
Can you add a test to `.github/workflows/ci.yml` testing that this compiles when `--features fips-3678` is set?

File: boring-sys/build.rs (hunk 3 above)
Comment (behrat) [reply boring-sys/build.rs:209]: In my build environment, the I must set the include path and library path to different values, or example:
```
$ env | grep BORING                                  
BORING_BSSL_INCLUDE_PATH=/opt/boringssl-fips/include
BORING_BSSL_LIB_PATH=/opt/boringssl-fips
```
Only then will it successfully build and link with my FIPS build of BoringSSL:
```
$ cargo run --features fips-3678 --example fips_enabled
   Compiling boring-sys v1.1.1 (/home/braden/rust/boring/boring-sys)
   Compiling boring v1.1.6 (/home/braden/rust/boring/boring)
    Finished dev [unoptimized + debuginfo] target(s) in 18.00s
     Running `/home/braden/rust/boring/target/debug/examples/fips_enabled`
boring::fips::enabled(): true
```
Open to any suggestions to simply things or make it work better in other use cases, though.

File: boring/src/x509/mod.rs (hunk 5 above)
Comment (jyn514) [reply boring/src/x509/mod.rs:240]: I think so, yes - that way the same code gets tested whether the feature is enabled or disabled. Maybe add a comment that this should be changed when the version of FIPS is bumped (I'm not sure what the process for that is; I guess the C boring lib itself has to get recertified?).

File: README.md (hunk 4 above)
Comment (jyn514) [reply README.md:38]: Hmm, ideally we could somehow give a compile error if you try and use a non-FIPS library when the FIPS feature is enabled. Maybe we can have it link to a symbol that only exists in the FIPS version? Or add an assert somewhere?

File: boring-sys/build.rs (hunk 1 above)
Comment (jyn514) [reply boring-sys/build.rs:179]: :+1: that would be great, thanks!

File: boring/src/x509/mod.rs (hunk 2 above)
Comment (jyn514) [reply boring/src/x509/mod.rs:511]: I think I was confused before - those docs are talking about `X509_getm_notAfter`, not `X509_get0_notAfter`. I think get0 should be ok as you're written it.

**Comment by jyn514** [comment]
I opened https://github.com/behrat/boring/pull/1 getting this working a little better without needing an external boring build.

File: boring-sys/build.rs [hunk 6]
```diff
@@ -216,10 +291,11 @@ fn main() {
...
 
-    let include_path = PathBuf::from(
-        std::env::var("BORING_BSSL_INCLUDE_PATH")
-            .unwrap_or_else(|_| String::from("deps/boringssl/src/include")),
-    );
+    let include_path = if cfg!(feature = "fips") {
+        format!("{}/include", BORING_SSL_PATH)
+    } else {
+        format!("{}/src/include", BORING_SSL_PATH)
+    };
```
Comment (jyn514) [inline boring-sys/build.rs:298]: I think I was confused before, `BORING_BSSL_INCLUDE_PATH` was supported even before this PR. The new code should look something like this:
```rust
    println!("cargo:rerun-if-env-changed=BORING_BSSL_INCLUDE_PATH");
    let include_path = std::env::var("BORING_BSSL_INCLUDE_PATH").unwrap_or_else(|| if cfg!(feature = "fips") {
        format!("{}/include", BORING_SSL_PATH)
    } else {
        format!("{}/src/include", BORING_SSL_PATH)
    });
```
Do you want to make that change or should I?

**Review by jyn514 — COMMENTED** [review]
Looks good with the `BORING_BSSL_INCLUDE_PATH` support added back in - sorry for the noise there.

**Comment by jyn514** [comment]
Oh, also it looks like you need to add clang-7 to the test environment.

File: boring/Cargo.toml [hunk 7]
```diff
@@ -21,3 +21,8 @@ boring-sys = { version = "1.1.0", path = "../boring-sys" }
 [dev-dependencies]
 hex = "0.4"
 rusty-hook = "^0.11"
+
+[features]
+fips-3678 = ["fips", "boring-sys/fips-3678"]
+# The "fips" feature is used for conditional-compilation that is common to all supported FIPS-validated versions
+fips = ["boring-sys/fips"]
```
[the remark this answers is not in the document]
Comment (behrat) [reply boring/Cargo.toml]: Discussed in https://github.com/behrat/boring/pull/1 and decided to only have `fips` feature for now (could support multiple versions later).

File: boring-sys/build.rs (hunk 6 above)
Comment (behrat) [reply boring-sys/build.rs:298]: Do you mind making the change? I promise to review quickly. :)

Comment (jyn514) [reply boring-sys/build.rs]: Pushed in https://github.com/cloudflare/boring/pull/52/commits/0e742156f58dc6ff4846efaefb4f95c17c5549ab.

**Comment by behrat** [comment]
Thanks for the help @jyn514! I think that addresses all feedback from my original version of the PR, and this looks ready-to-merge to me.

**Comment by wmorgan** [comment]
Thanks for this @jyn514 and @behrat!
